subsys/fs/nvs/nvs.c

```diff
@@ -6,104 +6,492 @@
+
+	while (len >= block_size) {
+		rc = _nvs_flash_read(fs, offset, buf, block_size);
+		if (rc) {
+			return rc;
+		}
+		rc = memcmp(data, buf, block_size);
+		if (rc) {
+			return rc;
+		}
+		len -= block_size;
+		offset += block_size;
+		data += block_size;
+	}
+	if (len) {
+		rc = _nvs_flash_read(fs, offset, buf, len);
+		if (rc) {
+			return rc;
+		}
+		rc = memcmp(data, buf, len);
+		if (rc) {
+			return rc;
+		}
+	}
+	return 0;
+}
+
+/* flash block move: move a block at offset to the current write location
+ * and updates the write location, len should be multiple of
+ * fs->write_block_size. _nvs_flash_move does a verification of the copy.
+ */
+int _nvs_flash_block_move(struct nvs_fs *fs, off_t offset, size_t len)
+{
+	int rc, byte_to_copy, block_size;
+	u8_t buf[max(NVS_BLOCK_SIZE, fs->write_block_size)];
+	off_t write_location;
+
+	write_location = fs->write_location;
+
+	block_size = sizeof(buf);
+
+	while (len > 0) {
+		byte_to_copy =	min(block_size, len);
+		rc = _nvs_flash_read(fs, offset, buf, byte_to_copy);
+
+		if (rc) {
+			return rc;
+		}
+		rc = _nvs_flash_write(fs, write_location, buf,
+				      byte_to_copy);
+		if (rc) {
+			return rc;
+		}
+		rc = _nvs_flash_block_cmp(fs, write_location, buf,
+					  byte_to_copy);
+		if (rc) {
+			return rc;
+		}
+		len -= byte_to_copy;
+		offset += byte_to_copy;
+		write_location += byte_to_copy;
+	}
+	/* update fs->write_location when all writes are successful */
+
+	fs->write_location = write_location;
+
+	return 0;
+}
+
+/* flash write and verify */
+int _nvs_flash_write_ver(struct nvs_fs *fs, off_t offset, const void *data,
+			 size_t len)
 {
-	struct _nvs_sector_hdr sector_hdr;
-	u16_t sector_hdr_len;
 	int rc;
 
-	/* just to be sure, align offset to sector_size */
-	offset &= ~(fs->sector_size - 1);
-	fs->sector_id += 1;
-	rc = _nvs_sector_hdr_get(fs, offset, &sector_hdr);
+	rc = _nvs_flash_write(fs, offset, data, len);
 	if (rc) {
-		SYS_LOG_DBG("Failed reading sector hdr");
 		return rc;
 	}
-	if (sector_hdr.fd_magic != 0xFFFFFFFF) {
-		SYS_LOG_DBG("Initializing non empty sector");
-		return -EIO;
+	rc = _nvs_flash_block_cmp(fs, offset, data, len);
+	if (rc) {
+		return rc;
 	}
-	sector_hdr.fd_magic = fs->magic;
-	sector_hdr.fd_id = fs->sector_id;
-	sector_hdr_len = _nvs_len_in_flash(fs, sizeof(struct _nvs_sector_hdr));
-	rc = nvs_flash_write(fs, offset, &sector_hdr, sizeof(sector_hdr));
+	return 0;
+}
+
+/* store an entry in flash at location offset */
+int _nvs_flash_wrt_entry(struct nvs_fs *fs, u16_t id, const void *data,
+			 size_t len)
+{
+	int rc;
+	off_t offset;
+	off_t fs_size = fs->sector_count * fs->sector_size;
+	size_t padding, flash_len, length_cnt, byte_cnt;
+	u8_t lenstr[2], ilenstr[2], buf[max(fs->write_block_size, 4)];
+
+	offset = fs->write_location;
+
+	/* create the length string to be written at start and end */
+	length_cnt = _nvs_put_length(lenstr, len);
+	if (length_cnt == 1) {
+		ilenstr[0] = lenstr[0];
+	} else {
+		ilenstr[0] = lenstr[1];
```

Comment:
Hi @nvlsianpu, thanks for all comments, the second length needs to be swapped because you will always read 2 bytes, but this can be data as well, so to make sure you get the correct length you are only sure about the last byte (this is definitely part of the length), so the last byte should be byte0 of the length.

Reply:
thank for elaboration.